### ISSUE TITLE
URL parsing security fix introduces regression

### ISSUE THREAD

**Issue by mmuehlenhoff** [facebook/hhvm#8104]
Hi,
the URL parsing fix (d163f8393a167552ec82c3b4a78bc4a9e7c8c157) which was also backported to the 3.23.4, 3.21.5, and 3.18.7 broke the Mediawiki test suite, please see 
https://phabricator.wikimedia.org/T185024 for details (and in particular https://phabricator.wikimedia.org/T185024#3904397 which pinpoints the problem).

**Comment by fredemmott** [comment]
Isolated repro:

```
var_dump(parse_url("http://xx23124:__ffdfdef__@www.test.com:12345/dir"));
```

Should not return false.

I'll reach out to our vendor about why that diff isn't a straight port of the PHP5 fix.

**Comment by fredemmott** [comment]
This appears to be caused by a change we applied, but I don't have context on why the approach was taken.

Most likely I'm going to throw away all our customizations to url_parse(), and re-sync with upstream instead.

**Comment by fredemmott** [comment]
For my reference: fix is D6798174 (not yet landed/on github)

**Comment by fredemmott** [comment]
Waiting for a clean internal test run before landing (some of our copies of PHP unit tests are out of sync) - https://gist.github.com/fredemmott/39c9abef4571f1e337d339fd8355da60 seems correct though (you'll need to apply patch with `-p2`)
